source/DataWriter.h

```diff
@@ -33,86 +35,125 @@ class DataNode;
+	DataWriter &Write(const DataNode &node);
 	// End the current line. This can be used to add line breaks or to terminate
 	// a line you have been writing token by token with WriteToken().
-	void Write();
+	DataWriter &Write();
 
 	// Begin a new line that is a "child" of the previous line.
-	void BeginChild();
+	DataWriter &BeginChild();
 	// Finish writing a block of child nodes and decrease the indentation.
-	void EndChild();
+	DataWriter &EndChild();
 
 	// Write a comment. It will be at the current indentation level, and will
 	// have "# " inserted before it.
-	void WriteComment(const std::string &str);
+	DataWriter &WriteComment(const std::string &str);
 
 	// Write a token, without writing a whole line. Use this very carefully.
-	void WriteToken(const char *a);
-	void WriteToken(const std::string &a);
+	DataWriter &WriteToken(const char *a);
+	DataWriter &WriteToken(const std::string &a);
 	// Write a token of any arithmetic type.
 	template <class A>
-	void WriteToken(const A &a);
+	DataWriter &WriteToken(const A &a);
+	// Writes a series of tokens without terminating the line.
+	template <class A, class ...B>
+	DataWriter &WriteToken(const A &a, B... others);
+	// Write the tokens of this DataNode without writing its children.
+	DataWriter &WriteTokens(const DataNode &node);
+
+	// Changes the separator used between tokens. The default is a single space.
+	DataWriter &SetSeparator(const std::string &sep);
+	// Changes the indentation used before child tokens. The default is a single tabulator.
+	DataWriter &SetIndentation(const std::string &indent);
 
 
 private:
-	// Save path (in UTF-8). Empty string for in-memory DataWriter.
-	std::string path;
 	// Current indentation level.
 	std::string indent;
+	// The string used for indentation.
+	std::string indentString = "\t";
 	// Before writing each token, we will write either the indentation string
 	// above, or this string.
-	static const std::string space;
+	std::string separator = " ";
 	// Remember which string should be written before the next token. This is
-	// "indent" for the first token in a line and "space" for subsequent tokens.
+	// "indent" for the first token in a line and "separator" for subsequent tokens.
+	// Null pointer signals that this string was already written for the next token.
 	const std::string *before;
 	// Compose the output in memory before writing it to file.
 	std::ostringstream out;
 };
 
 
 
+// Raw write for numeric types.
+template <class C>
+DataWriter &DataWriter::WriteRaw(const C &c)
+{
+	static_assert(std::is_arithmetic<C>::value,
+		"DataWriter cannot output anything but strings and arithmetic types.");
+	out << c;
+	return *this;
+}
+
+
+
 // The Write() function can take any number of arguments, each of which becomes
 // a token. They must be either strings or numeric types.
 template <class A, class ...B>
-void DataWriter::Write(const A &a, B... others)
+DataWriter &DataWriter::Write(const A &a, B... others)
```

Comment:
Why does the original code not take `B`s by `const&`? It really should to avoid all those copies.

Reply:
It's giving me compiler errors... how am I supposed to write that?

I tried `const B... &others` and `const &B... others` but neither seems to work.

Reply:
`const B &...others`